nautobot/extras/tests/test_dynamicgroups.py

```diff
@@ -0,0 +1,360 @@
+                name="Site 1",
+                slug="site-1",
+                filter={"site": ["site-1"]},
+                content_type=cls.device_ct,
+            ),
+            # No matches (bogues name match)
+            DynamicGroup.objects.create(
+                name="No Filter",
+                slug="no-filter",
+                description="A group with no filter",
+                filter={"name": ["bogus"]},
+                content_type=cls.device_ct,
+            ),
+            # Site-2 only
+            DynamicGroup.objects.create(
+                name="Site 3",
+                slug="site-3",
+                filter={"site": ["site-3"]},
+                content_type=cls.device_ct,
+            ),
+        ]
+
+
+class DynamicGroupModelTest(DynamicGroupTestBase):
+    def test_content_type_is_immutable(self):
+        """Test that `content_type` is immutable after create."""
+        instance = self.groups[0]
+        with self.assertRaises(ValidationError):
+            instance.content_type = self.dynamicgroup_ct
+            instance.validated_save()
+
+    def test_clean_filter_not_dict(self):
+        """Test that invalid filter types raise errors."""
+        instance = self.groups[0]
+        with self.assertRaises(ValidationError):
+            instance.filter = None
+            instance.validated_save()
+
+        with self.assertRaises(ValidationError):
+            instance.filter = []
+            instance.validated_save()
+
+        with self.assertRaises(ValidationError):
+            instance.filter = "site=ams01"
+            instance.validated_save()
+
+    def test_clean_filter_not_valid(self):
+        """Test that an invalid filter dict raises an error."""
+        instance = self.groups[0]
+        with self.assertRaises(ValidationError):
+            instance.filter = {"site": -42}
+            instance.validated_save()
+
+    def test_clean_valid(self):
+        """Test a clean validation."""
+        group = self.groups[0]
+        group.refresh_from_db()
+        old_filter = group.filter
+
+        # Overload the filter and validate that it is the same afterward.
+        new_filter = {"interfaces": True}
+        group.set_filter(new_filter)
+        group.validated_save()
+        self.assertEqual(group.filter, new_filter)
+
+        # Restore the old filter.
+        group.filter = old_filter
+        group.save()
+
+    def test_get_for_object(self):
+        """Test `DynamicGroup.objects.get_for_object()`."""
+        device1 = self.devices[0]  # site-1
+        device2 = self.devices[1]  # site-2
+
+        # Assert that the groups we got from `get_for_object()` match the lookup
+        # from the group instance itself.
+        device1_groups = DynamicGroup.objects.get_for_object(device1)
+        self.assertEqual(
+            sorted(device1_groups.values_list("pk", flat=True)),
+            sorted(device1.dynamic_groups.values_list("pk", flat=True)),
+        )
```

Comment:
Might be good to explicitly assert the value of either of these `values_list()` returns, rather than just asserting that they're equal, whatever they might be?

Reply:
When it comes to lists, membership values being identical is equality. In this case in particular it's a list of UUID objects, so I feel it's a sane assertion.

Reply:
My point was that this test will pass as written even if both `values_list()` calls return `["Surprise!"]`, because we don't care what's actually in these lists, just that they happen to be equal. I'd like to think we could compare against the actual expected value - in this case I think it'd be something like `[self.groups[0].pk]`?